net: samples: zperf: Refactor to use socket API

rlubos opened the pull request:
Refactor `zperf` sample to use socket API instead of `net_context` directly. Additionally, submit 2 fixes that help to retain performance when socket API is used:
* ~~Add `k_yield()` in the RX thread to allow other threads to run when bulk of packets is queued for processing.~~
* Avoid unnecessary packet cloning on the TCP RX path

I can split the fixes into individual PRs if requested.

rlubos committed:
samples: net: zperf: Rewrite upload part to use sockets

Rewrite TCP/UDP upload part of the zperf sample to use socket API
instead of net_context. This has a negligible impact on the upload
throughputs (< 1 Mbps).

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
samples: net: zperf: Rewrite download part to use sockets

Rewrite the TCP/UDP download part of the zperf sample to use socket API.
For UDP, performance impact is negligible (< 1 Mbps), for TCP it's
noticable, but still throughputs can be considered satisfactory (up to
~75 Mbps).

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

rlubos committed:
net: tcp: Avoid unnecessary packet cloning on the RX path

So far, TCP cloned a packet with data on an RX path for the application,
leaving the original packet intact. This isn't really needed, as the
original packet is unconditionally freed later anyway, so the TCP can as
well simply queue the original packet for the application, while
informing the network processing core, that the packet was consumed by
the TCP layer.

This allows to improve the download throughput even further, since the
CPU don't waste time on needles packet copying.

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>

erwango commented:
> @erwango I've noticed that with `eth_stm32_hal.c` driver I can get even better download results if I enable HW checksum calculation ([ETH_CHECKSUM_BY_HARDWARE](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/ethernet/eth_stm32_hal.c#L1096)) and add RX checksum computation to the capabilities ([ETHERNET_HW_RX_CHKSUM_OFFLOAD](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/ethernet/eth_stm32_hal.c#L968)) - up to 85 MBit/sec for TCP download - I really doubt we could get any better numbers for 100Mbit Ethernet.
> 
> I'm not familiar enough with the driver to tell if those changes are enough, or if there are any other consequences that I don't see - do you think it's ok to make such changes in the upstream driver?

That's a good result indeed.
I think this configuration is default since the driver was first upstreamed. I'm not sure about the drawback to convert to these settings. What can be done would be to add this as a configuration option, defaulting to this news values. People seeing adverse effects will be able to report and switch back to the previous settings.

rlubos committed:
samples: net: zperf: Comply with Coccinelle requirements

Cocinelle complains about "shell" being used as a parameter name:

"Violation to rule 5.7 (Tag name should be unique) tag: shell"

Therefore rename the parameter throughout the sample to "sh".

Signed-off-by: Robert Lubos <robert.lubos@nordicsemi.no>
